File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int solve(TreeNode* root, int& maxi){
        if(!root) return 0;
        int lh = solve(root->left, maxi);
        int rh = solve(root->right, maxi);
        maxi = max(maxi, root->val + lh + rh);
        return max(root->val+ max(rh,lh), 0);
    }
    int maxPathSum(TreeNode* root) {
        int maxi = root->val;
        solve(root, maxi);
        return maxi;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.cnt = 0
        def solve(node, maxVal):
            if not node:
                return
            if node.val >= maxVal:
                self.cnt += 1
                maxVal = node.val
            solve(node.left, maxVal)
            solve(node.right, maxVal)
        solve(root, float("-inf"))
        return self.cnt